posts/2019-01-14-easy-peasy-nix-versions.md

```diff
@@ -0,0 +1,175 @@
+single package for simplicity's sake):
+
+``` bash
+versions=...    # The file `versions.json`
+package=...     # The package to update
+```
+
+Then it reads that package's owner, repository name and the branch that we're
+tracking:
+
+``` bash
+owner=$(cat $versions | jq -r ".[\"$package\"].owner")
+repo=$(cat $versions | jq -r ".[\"$package\"].repo")
+branch=$(cat $versions | jq -r ".[\"$package\"].branch")
+```
+
+Using this it retrieves the GitHub URL and calls `nix-prefetch-url` to compute
+the sha256:
+
+``` bash
+# Ask GitHub what the latest commit (revision) is on $branch:
+new_rev=$(curl -sfL \
+            https://api.github.com/repos/$owner/$repo/git/refs/heads/$branch \
+            | jq -r .object.sha)
+# Craft the URL of that commit's archive:
+url=https://github.com/$owner/$repo/archive/$new_rev.tar.gz
+# Compute the archive's sha256:
+new_sha256=$(nix-prefetch-url --unpack "$url")
+```
+
+Finally it re-reads the `versions.json` content, updates the `rev` and `sha256`
+attributes of that package, and writes it back to `versions.json`:
+
+``` bash
+res=$(cat $versions \
+    | jq -rM ".[\"$package\"].rev = \"$new_rev\"" \
+    | jq -rM ".[\"$package\"].sha256 = \"$new_sha256\""
+    )
+
+echo "$res" > $versions
+```
+
+Package: UPDATED!
+
+## Closing thoughts
+
+As mentioned in the introduction, this is a convention and should be tweaked to
+fit your needs. The important points are:
+
+* All versions and package specs should live in a single file.
+* The specs should be machine readable for automated update.
+
+When I create a new Nix project I simply copy over the [`nix/fetch.nix`] file
+and the [update script][`script/update`]. This cannot be abstracted away in a
+separate repo because there's a chicken and egg problem: if this is a Nix
+"library", how do we fetch it?
+
+The `versions.json` file is very simple; YMMV. For instance you could store
+URLs instead of GitHub repositories (you should then also tweak the update
+script). The [`script/update`] can also be adapted to your needs; for instance,
+in [homies], it's possible to only compute the `sha256` without pulling the
+branch's latest revision. That's super convenient when adding new packages: you
+don't have to call `nix-prefetch-url` by hand!
+
+[`default.nix`]: https://github.com/nmattia/homies/blob/06aa54743990613f3b53a3bf7334d23ce59acc4a/default.nix
+[`jq`]: https://stedolan.github.io/jq/
+[`nix/fetch.nix`]: https://github.com/nmattia/homies/blob/06aa54743990613f3b53a3bf7334d23ce59acc4a/nix/fetch.nix
+[`nix/versions.json`]: https://github.com/nmattia/homies/blob/06aa54743990613f3b53a3bf7334d23ce59acc4a/nix/versions.json
+[`script/update`]: https://github.com/nmattia/homies/blob/06aa54743990613f3b53a3bf7334d23ce59acc4a/script/update
+[homies-article]: https://nmattia.com/posts/2018-03-21-nix-reproducible-setup-linux-macos.html
+[homies]: https://github.com/nmattia/homies
+[require.nix]: https://www.youtube.com/watch?v=DHOLjsyXPtM
```

Comment:
this has actually been spun off and is now called [flakes.nix](https://github.com/nix-community/flake.nix/)